File: data/scripts/controllers/projectile.lua
local C = {}

function C:init(unit)
  self.unit = unit
end

function C:onUpdate()
  if not self.unit.root then return end
  self.unit.root.position = self.unit.root.position + self.unit.velocity:mulScalarReturn(self.unit.speed * game.deltaTime)
  self.unit.speed = self.unit.speed + self.unit.speed * self.unit.acceleration * game.deltaTime
  util.clampValue(self.unit.speed, self.unit.minSpeed, self.unit.maxSpeed)
end

function C:onSerialize(data)
  data.unitId = self.unit.id
end

function C:onDeserialize(data)
  self.unit = unitFromId(data.unitId)
end

return newInstance(C)

File: src/engine/projectile.cpp
#include "projectile.h"
#include "sprite.h"
#include "unit.h"
#include "game.h"
#include "resources/unit_resource.h"
#include "resources/script_resource.h"

namespace engine
{
void Projectile::copyFrom(struct Unit* other)
{
	Unit::copyFrom(other);
	scriptClass = other->unitResource->script ? other->unitResource->script->createClassInstance(this) : nullptr;
}

void Projectile::initializeFrom(struct UnitResource* res)
{
	Unit::initializeFrom(res);
	scriptClass = res->script ? res->script->createClassInstance(this) : nullptr;
}

void Projectile::update(Game* game)
{
	Unit::update(game);

	f32 deltaTime = isPlayer() ? game->realDeltaTime : game->deltaTime;

	if (!root) return;
	root->position += velocity * speed * deltaTime;
	speed += speed * acceleration * deltaTime;
	clampValue(speed, minSpeed, maxSpeed);
}

}
